### PR TITLE
Restructure differentiation schedule into a breadth first traversal

### PR DESCRIPTION
**Pull request by vaithak**

### CODE REVIEW

File: lib/Differentiator/DiffPlanner.cpp [hunk 1]
```diff
@@ -275,8 +275,25 @@ namespace clad {
   void DiffRequest::UpdateDiffParamsInfo(Sema& semaRef) {
     // Diff info for pullbacks is generated automatically,
     // its parameters are not provided by the user.
-    if (Mode == DiffMode::experimental_pullback)
+    if (Mode == DiffMode::experimental_pullback) {
+      if (!Function->getPreviousDecl())
```
Comment (vgvassilev) [inline lib/Differentiator/DiffPlanner.cpp:279]: What's the intent of this early exit here? Can we have a comment?

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 2]
```diff
@@ -386,57 +389,62 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+        // Creates the ArraySubscriptExprs for the independent variables
+        size_t idx = 0;
+        for (const auto* arg : args) {
+          // FIXME: fix when adding array inputs, now we are just skipping all
+          // array/pointer inputs (not treating them as independent variables).
+          if (utils::isArrayOrPointerType(arg->getType())) {
+            if (arg->getName() == "p")
+              m_Variables[arg] = m_Result;
+            idx += 1;
+            continue;
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:407]: Can we come up with a test here?

**Review by vgvassilev — COMMENTED** [review]
Thank you! This is a major improvement. We have discussed that now we can process diff schedules separately from the differentiation process. That is, we first plan and then act. This PR does not get us there just yet, and this is why it relies on a forward declarations because it only appends to the diff plan instead of inserting.

That's an incremental change but we should also try to make the next step maybe in a separate PR.

File: lib/Differentiator/DiffPlanner.cpp (hunk 1 above)
Comment (vaithak) [reply lib/Differentiator/DiffPlanner.cpp:281]: added a comment 👍🏼

File: include/clad/Differentiator/DerivativeBuilder.h [hunk 3]
```diff
@@ -37,6 +37,7 @@ namespace clad {
     class CladPlugin;
     clang::FunctionDecl* ProcessDiffRequest(CladPlugin& P,
                                             DiffRequest& request);
+    void AddRequestToSchedule(CladPlugin& P, const DiffRequest& request);
```
[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/DerivativeBuilder.h:42]: added a fixme 👍🏼

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 2 above)
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:407]: This is for Jacobian computation and the particular case is for something related to ROOT as captured here: https://github.com/vgvassilev/clad/pull/478#discussion_r944066696.
I will look into this when I work on the Jacobain array-related issue: https://github.com/vgvassilev/clad/issues/472

File: include/clad/Differentiator/DerivativeBuilder.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/DerivativeBuilder.h:42]: Since this is more of a design issue we should keep track of it in the issue tracker to avoid dropping from the priority list.

[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/DerivativeBuilder.h:42]: https://github.com/vgvassilev/clad/issues/857

**Review by parth-07 — APPROVED** [review]
Looks really good.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -911,7 +911,7 @@ jobs:
         cat obj/CMakeCache.txt
         cat obj/CMakeFiles/*.log
     - name: Setup tmate session
-      if: ${{ failure() && runner.debug }}
+      if: ${{ failure() && matrix.debug_build == true }}
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:914]: Can we undo this change?

**Review by vgvassilev — COMMENTED** [review]
Looks like the clang-tidy reports are relevant. Can you address them?

File: .github/workflows/ci.yml (hunk 4 above)
Comment (vaithak) [reply .github/workflows/ci.yml]: removed 👍🏼

File: lib/Differentiator/VisitorBase.cpp [hunk 5]
```diff
@@ -524,13 +528,17 @@ namespace clad {
...
+    if (Loc.isInvalid())
+      Loc = m_Function->getLocation();
     UnqualifiedId Member;
-    Member.setIdentifier(&m_Context.Idents.get(MemberFunctionName), noLoc);
+    Member.setIdentifier(&m_Context.Idents.get(MemberFunctionName), Loc);
     CXXScopeSpec SS;
     bool isArrow = Base->getType()->isPointerType();
     auto ME = m_Sema
-                  .ActOnMemberAccessExpr(getCurrentScope(), Base, noLoc,
+                  .ActOnMemberAccessExpr(getCurrentScope(), Base, Loc,
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/VisitorBase.cpp:540]: This one seems irrelevant, it is using some heuristic to deduce that `noLoc` is more similar to `OpLoc`, maybe because of Levenshtein distance as mentioned here: https://releases.llvm.org/13.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/readability-suspicious-call-argument.html#levenshtein-distance-as-levenshtein.

**Review by vgvassilev — APPROVED** [review]
LGTM! Do you want the merge the 6 commits the way they are or you want to squash or something else?

**Comment by vaithak** [comment]
I think the current order of commits is fine.
